File: src/com/leo/leetcode/algorithm/q0400/Q451.java
package com.leo.leetcode.algorithm.q0400;

import java.util.*;

/**
 * 给定一个字符串，请将字符串里的字符按照出现的频率降序排列。
 * 链接：https://leetcode-cn.com/problems/sort-characters-by-frequency/
 */
public class Q451 {
    public static void main(String[] args) {
        // "eert"
        System.out.println(new Q451().frequencySort("tree"));
        // "aaaccc"
        System.out.println(new Q451().frequencySort("cccaaa"));
        // "bbAa"
        System.out.println(new Q451().frequencySort("Aabb"));
    }

    public String frequencySort(String s) {
        Map<Character, Integer> m = new HashMap<>();
        StringBuilder ret = new StringBuilder();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            m.put(c, m.getOrDefault(c, 0) + 1);
        }
        PriorityQueue<int[]> q = new PriorityQueue<>((o1, o2) -> o2[1] - o1[1]);
        for (Map.Entry<Character, Integer> it : m.entrySet()) {
            q.add(new int[]{it.getKey(), it.getValue()});
        }
        while (!q.isEmpty()) {
            int[] p = q.poll();
            while (p[1]-- > 0) {
                ret.append((char) p[0]);
            }
        }
        return ret.toString();
    }
}

File: py/q0400/Q451.py
"""
 * 给定一个字符串 s ，根据字符出现的 频率 对其进行 降序排序 。一个字符出现的 频率 是它出现在字符串中的次数。
 * 返回 已排序的字符串 。如果有多个答案，返回其中任何一个。
 * 提示:
 * 1、1 <= s.length <= 5 * 10^5
 * 2、s 由大小写英文字母和数字组成
 * 链接：https://leetcode.cn/problems/sort-characters-by-frequency
"""
from typing import Counter


class Solution:

    def frequencySort(self, s: str) -> str:
        arr = sorted([[v, k] for k, v in Counter(s).items()], reverse=True)
        ans = [k * v for v, k in arr]
        return ''.join(ans)


if __name__ == '__main__':
    # "eert"
    print(Solution().frequencySort("tree"))
    # "aaaccc"
    print(Solution().frequencySort("cccaaa"))
    # "bbAa"
    print(Solution().frequencySort("Aabb"))